sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala

```diff
@@ -232,7 +232,9 @@ private[csv] object CSVTypeCast {
       nullable: Boolean = true,
       options: CSVOptions = CSVOptions()): Any = {
 
-    if (nullable && datum == options.nullValue) {
+    // datum can be null if the number of fields found is less than the length of the schema
+    if (datum == options.nullValue || datum == null) {
+      require(nullable, "null value found but the field is not nullable.")
```

Comment:
(Other related PRs with the comment, https://github.com/apache/spark/pull/15767#discussion_r86675671, are  https://github.com/apache/spark/pull/15329  and https://github.com/apache/spark/pull/14124)